sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -735,4 +745,100 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
         df1.groupBy("j").agg(max("k")))
     }
   }
+
+  //  A test with a partition where the number of files in the partition is
+  //  large. tests for the condition where the serialization of such a task may result in a stack
+  //  overflow if the files list is stored in a recursive data structure
+  test("SPARK-27100 stack overflow: read data with large partitions") {
+    // Need a large number of files in the partition for the overflow
+    val numFilesInPartition = 100000
+
+    def getInputRDD(sparkSession: SparkSession, isBucketed: Boolean): RDD[InternalRow] = {
+      val partitionValues = InternalRow.apply(Array("a"))
+
+      val schema = new StructType()
+      val fakeHadoopFsRelation = new HadoopFsRelation(null, schema, schema, null,
+        new ParquetFileFormat, null)(sparkSession)
+      val optionalBucketSet = null
+      val bucketSpec = new BucketSpec(1, Seq("a"), Seq("b"))
+      val files = (0 to numFilesInPartition - 1).toStream.map { i =>
+        new FileStatus(10, false, 1, 512, 1000,
+          new Path(s"file${i}_0.zzz"))
+      }
+      val partitionDirectory = PartitionDirectory(partitionValues, files);
+
+      val fileSource =
+        FileSourceScanExec(fakeHadoopFsRelation,
+          null,
+          schema,
+          null,
+          Option(optionalBucketSet),
+          Seq.empty,
+          Option(new TableIdentifier("stackOverflow")))
+
+      val inputRDD = if (isBucketed) {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createBucketedReadRDD)
+
+        fileSource invokePrivate createBucketedReadRDD(bucketSpec,
+          (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+          Array(partitionDirectory),
+          fakeHadoopFsRelation)
+      } else {
+        // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+        val createNonBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createNonBucketedReadRDD)
+
+        fileSource invokePrivate createNonBucketedReadRDD(
```

Comment:
We typically don't use this style. `fileSource.invokePrivate.createNonBucketedReadRDD`